File: tests/py/test_frame_format_serialization.py
#!/usr/bin/env python3
"""
Frame format serialization test for Python.

This test serializes data using different frame formats and saves to binary files.
Usage: test_frame_format_serialization.py <frame_format>
"""

import sys
import os
import json


def print_failure_details(label, expected_values=None, actual_values=None, raw_data=None):
    """Print detailed failure information"""
    print(f"\n{'='*60}")
    print(f"FAILURE DETAILS: {label}")
    print(f"{'='*60}")
    
    if expected_values:
        print("\nExpected Values:")
        for key, val in expected_values.items():
            print(f"  {key}: {val}")
    
    if actual_values:
        print("\nActual Values:")
        for key, val in actual_values.items():
            print(f"  {key}: {val}")
    
    if raw_data:
        print(f"\nRaw Data ({len(raw_data)} bytes):")
        print(f"  Hex: {raw_data.hex()}")
    
    print(f"{'='*60}\n")


def load_expected_values():
    """Load expected values from JSON file"""
    json_path = os.path.join(os.path.dirname(__file__), '..', 'expected_values.json')
    try:
        with open(json_path, 'r') as f:
            data = json.load(f)
        return data['serialization_test']
    except Exception as e:
        print(f"Error loading expected values: {e}")
        return None


def get_frame_parser_class(frame_format):
    """Get the frame parser class for a given frame format name"""
    import importlib
    
    # Map frame format names to class names
    format_class_map = {
        'basic_default': 'BasicDefault',
        'basic_minimal': 'BasicMinimal',
        'basic_seq': 'BasicSeq',
        'basic_sys_comp': 'BasicSysComp',
        'basic_extended_msg_ids': 'BasicExtendedMsgIds',
        'basic_extended_length': 'BasicExtendedLength',
        'basic_extended': 'BasicExtended',
        'basic_multi_system_stream': 'BasicMultiSystemStream',
        'basic_extended_multi_system_stream': 'BasicExtendedMultiSystemStream',
        'tiny_default': 'TinyDefault',
        'tiny_minimal': 'TinyMinimal',
        'tiny_seq': 'TinySeq',
        'tiny_sys_comp': 'TinySysComp',
        'tiny_extended_msg_ids': 'TinyExtendedMsgIds',
        'tiny_extended_length': 'TinyExtendedLength',
        'tiny_extended': 'TinyExtended',
        'tiny_multi_system_stream': 'TinyMultiSystemStream',
        'tiny_extended_multi_system_stream': 'TinyExtendedMultiSystemStream',
    }
    
    class_name = format_class_map.get(frame_format)
    if not class_name:
        return None, None
    
    # Import the module dynamically using importlib
    # The generated code uses relative imports, so import through the package
    # The PYTHONPATH should be set to tests/generated (not tests/generated/py)
    try:
        # First try importing from py package
        module = importlib.import_module(f"py.{frame_format}")
        parser_class = getattr(module, class_name)
        return parser_class, class_name
    except (ImportError, AttributeError):
        pass
    
    # Fallback: try direct import (if PYTHONPATH includes generated/py directly)
    try:
        module = importlib.import_module(frame_format)
        parser_class = getattr(module, class_name)
        return parser_class, class_name
    except (ImportError, AttributeError) as e:
        print(f"  Error loading frame format {frame_format}: {e}")
        return None, None


def create_test_data(frame_format):
    """Create test data for the specified frame format"""
    try:
        # Import the message class (works with PYTHONPATH including generated parent)
        import importlib
        try:
            msg_module = importlib.import_module('py.serialization_test_sf')
            SerializationTestSerializationTestMessage = msg_module.SerializationTestSerializationTestMessage
        except ImportError:
            try:
                from serialization_test_sf import SerializationTestSerializationTestMessage
            except ImportError as e:
                print(f"  Error importing serialization_test_sf: {e}")
                return False

        # Get frame parser class
        parser_class, class_name = get_frame_parser_class(frame_format)
        if parser_class is None:
            print(f"  Unknown frame format: {frame_format}")
            return False

        # Load expected values from JSON
        expected = load_expected_values()
        if not expected:
            return False

        # Create test message using values from JSON
        msg = SerializationTestSerializationTestMessage(
            magic_number=expected['magic_number'],
            test_string=expected['test_string'].encode('utf-8'),
            test_float=expected['test_float'],
            test_bool=expected['test_bool'],
            test_array=expected['test_array']
        )

        # Create parser instance and encode using the frame format
        parser = parser_class()
        encoded_data = parser.encode_msg(msg)

        # Write to file
        filename = f'python_{frame_format}_test_data.bin'
        with open(filename, 'wb') as f:
            f.write(bytes(encoded_data))

        print(f"  [OK] Encoded with {frame_format} format ({len(encoded_data)} bytes) -> {filename}")
        return True

    except ImportError as e:
        print(f"  Import error: {e}")
        return True  # Skip if generated code not available

    except Exception as e:
        print_failure_details(
            f"Create test data exception: {type(e).__name__}",
            expected_values={"result": "success"},
            actual_values={"exception": str(e)}
        )
        import traceback
        traceback.print_exc()
        return False


def get_supported_formats():
    """Return list of supported frame formats"""
    return [
        'basic_default', 'basic_minimal', 'basic_seq', 'basic_sys_comp',
        'basic_extended_msg_ids', 'basic_extended_length', 'basic_extended',
        'basic_multi_system_stream', 'basic_extended_multi_system_stream',
        'tiny_default', 'tiny_minimal', 'tiny_seq', 'tiny_sys_comp',
        'tiny_extended_msg_ids', 'tiny_extended_length', 'tiny_extended',
        'tiny_multi_system_stream', 'tiny_extended_multi_system_stream',
    ]


def main():
    """Main test function"""
    print("\n[TEST START] Python Frame Format Serialization")

    if len(sys.argv) != 2:
        print(f"  Usage: {sys.argv[0]} <frame_format>")
        print("  Supported formats:")
        for fmt in get_supported_formats():
            print(f"    {fmt}")
        print("[TEST END] Python Frame Format Serialization: FAIL\n")
        return False
    
    frame_format = sys.argv[1]
    success = create_test_data(frame_format)
    
    status = "PASS" if success else "FAIL"
    print(f"[TEST END] Python Frame Format Serialization: {status}\n")
    
    return success


if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: tests/ts/test_frame_format_serialization.ts
/**
 * Frame format serialization test for TypeScript.
 * 
 * This test serializes data using different frame formats and saves to binary files.
 * Usage: test_frame_format_serialization.ts <frame_format>
 */
import * as fs from 'fs';
import * as path from 'path';

function printFailureDetails(label: string, expectedValues?: any, actualValues?: any, rawData?: Buffer): void {
  console.log('\n============================================================');
  console.log(`FAILURE DETAILS: ${label}`);
  console.log('============================================================');
  
  if (expectedValues) {
    console.log('\nExpected Values:');
    for (const [key, val] of Object.entries(expectedValues)) {
      console.log(`  ${key}: ${val}`);
    }
  }
  
  if (actualValues) {
    console.log('\nActual Values:');
    for (const [key, val] of Object.entries(actualValues)) {
      console.log(`  ${key}: ${val}`);
    }
  }
  
  if (rawData && rawData.length > 0) {
    console.log(`\nRaw Data (${rawData.length} bytes):`);
    console.log(`  Hex: ${rawData.toString('hex').substring(0, 128)}${rawData.length > 64 ? '...' : ''}`);
  }
  
  console.log('============================================================\n');
}

function loadExpectedValues(): any {
  try {
    const jsonPath = path.join(__dirname, '../../../expected_values.json');
    const data = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));
    return data.serialization_test;
  } catch (error) {
    console.log(`Error loading expected values: ${error}`);
    return null;
  }
}

interface FrameFormatConfig {
  start: number[];
  headerSize: number;
  footerSize: number;
  hasLength: boolean;
  lengthBytes: number;
}

// Frame format configurations
const FRAME_FORMATS: { [key: string]: FrameFormatConfig } = {
  'basic_default': { start: [0x90, 0x71], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_minimal': { start: [0x90, 0x70], headerSize: 3, footerSize: 0, hasLength: false, lengthBytes: 0 },
  'basic_seq': { start: [0x90, 0x76], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_sys_comp': { start: [0x90, 0x75], headerSize: 6, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_extended_msg_ids': { start: [0x90, 0x72], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_extended_length': { start: [0x90, 0x73], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'basic_extended': { start: [0x90, 0x74], headerSize: 6, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'basic_multi_system_stream': { start: [0x90, 0x77], headerSize: 7, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'basic_extended_multi_system_stream': { start: [0x90, 0x78], headerSize: 9, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'tiny_default': { start: [0x71], headerSize: 3, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_minimal': { start: [0x70], headerSize: 2, footerSize: 0, hasLength: false, lengthBytes: 0 },
  'tiny_seq': { start: [0x76], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_sys_comp': { start: [0x75], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_extended_msg_ids': { start: [0x72], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_extended_length': { start: [0x73], headerSize: 4, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'tiny_extended': { start: [0x74], headerSize: 5, footerSize: 2, hasLength: true, lengthBytes: 2 },
  'tiny_multi_system_stream': { start: [0x77], headerSize: 6, footerSize: 2, hasLength: true, lengthBytes: 1 },
  'tiny_extended_multi_system_stream': { start: [0x78], headerSize: 8, footerSize: 2, hasLength: true, lengthBytes: 2 },
};

function createTestData(frameFormat: string): boolean {
  try {
    const config = FRAME_FORMATS[frameFormat];
    if (!config) {
      console.log(`  Unknown frame format: ${frameFormat}`);
      return false;
    }

    // Load expected values from JSON
    const expected = loadExpectedValues();
    if (!expected) {
      return false;
    }

    const msg_id = 204;
    
    // Create payload matching the message structure
    const payloadSize = 95;
    const payload = Buffer.alloc(payloadSize);
    let offset = 0;
    
    // magic_number (uint32, little-endian)
    payload.writeUInt32LE(expected.magic_number, offset);
    offset += 4;
    
    // test_string: length byte + 64 bytes of data
    const testString = expected.test_string;
    payload.writeUInt8(testString.length, offset);
    offset += 1;
    Buffer.from(testString).copy(payload, offset, 0, testString.length);
    offset += 64;
    
    // test_float (float, little-endian)
    payload.writeFloatLE(expected.test_float, offset);
    offset += 4;
    
    // test_bool (1 byte)
    payload.writeUInt8(expected.test_bool ? 1 : 0, offset);
    offset += 1;
    
    // test_array: count byte + int32 data (5 elements max)
    const testArray: number[] = expected.test_array;
    payload.writeUInt8(testArray.length, offset);
    offset += 1;
    for (let i = 0; i < 5; i++) {
      if (i < testArray.length) {
        payload.writeInt32LE(testArray[i], offset);
      } else {
        payload.writeInt32LE(0, offset);
      }
      offset += 4;
    }
    
    // Build frame based on format
    const totalSize = config.headerSize + payloadSize + config.footerSize;
    const frame = Buffer.alloc(totalSize);
    let frameOffset = 0;
    
    // Start bytes
    for (const b of config.start) {
      frame[frameOffset++] = b;
    }
    
    // Length field (if present)
    if (config.hasLength) {
      if (config.lengthBytes === 2) {
        frame.writeUInt16LE(payloadSize, frameOffset);
        frameOffset += 2;
      } else {
        frame[frameOffset++] = payloadSize & 0xFF;
      }
    }
    
    // Msg ID
    frame[frameOffset++] = msg_id;
    
    // Payload
    payload.copy(frame, frameOffset);
    frameOffset += payloadSize;
    
    // CRC (if present)
    if (config.footerSize > 0) {
      // Calculate Fletcher checksum starting after start bytes
      let byte1 = 0;
      let byte2 = 0;
      for (let i = config.start.length; i < frameOffset; i++) {
        byte1 = (byte1 + frame[i]) & 0xFF;
        byte2 = (byte2 + byte1) & 0xFF;
      }
      frame[frameOffset++] = byte1;
      frame[frameOffset++] = byte2;
    }
    
    // Write to file
    const outputPath = fs.existsSync('tests/generated/ts/js') 
      ? `tests/generated/ts/js/typescript_${frameFormat}_test_data.bin`
      : `typescript_${frameFormat}_test_data.bin`;
    fs.writeFileSync(outputPath, frame);

    console.log(`  [OK] Encoded with ${frameFormat} format (${frame.length} bytes) -> ${outputPath}`);
    return true;
  } catch (error) {
    printFailureDetails(`Create test data exception: ${error}`);
    return false;
  }
}

function main(): boolean {
  console.log('\n[TEST START] TypeScript Frame Format Serialization');
  
  const args = process.argv.slice(2);
  if (args.length !== 1) {
    console.log(`  Usage: ${process.argv[1]} <frame_format>`);
    console.log('  Supported formats:');
    for (const fmt of Object.keys(FRAME_FORMATS)) {
      console.log(`    ${fmt}`);
    }
    console.log('[TEST END] TypeScript Frame Format Serialization: FAIL\n');
    return false;
  }
  
  try {
    if (!createTestData(args[0])) {
      console.log('[TEST END] TypeScript Frame Format Serialization: FAIL\n');
      return false;
    }

    console.log('[TEST END] TypeScript Frame Format Serialization: PASS\n');
    return true;
  } catch (error) {
    printFailureDetails(`Exception: ${error}`);
    console.log('[TEST END] TypeScript Frame Format Serialization: FAIL\n');
    return false;
  }
}

if (require.main === module) {
  const success = main();
  process.exit(success ? 0 : 1);
}

export { main };
